org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchExpression.java

```diff
@@ -341,31 +344,26 @@ public TypeBinding resolveType(BlockScope upperScope) {
 					return this.resolvedType = null;
 				}
 
-				if (this.originalValueResultExpressionTypes == null) {
-					this.originalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
-					this.finalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
-					for (int i = 0; i < resultExpressionsCount; ++i) {
-						this.finalValueResultExpressionTypes[i] = this.originalValueResultExpressionTypes[i] =
-								this.resultExpressions.get(i).resolvedType;
-					}
+				this.originalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
+				this.finalValueResultExpressionTypes = new TypeBinding[resultExpressionsCount];
+				for (int i = 0; i < resultExpressionsCount; ++i) {
+					this.finalValueResultExpressionTypes[i] = this.originalValueResultExpressionTypes[i] = this.resultExpressions.get(i).resolvedType;
 				}
+
 				if (isPolyExpression()) { //The type of a poly switch expression is the same as its target type.
 					if (this.expectedType == null || !this.expectedType.isProperType(true)) {
 						return new PolyTypeBinding(this);
 					}
 					return this.resolvedType = computeConversions(this.scope, this.expectedType) ? this.expectedType : null;
 				}
-			} else {
-				// re-resolving of poly expression:
+			} else { // re-resolving of poly expression:
 				resultExpressionsCount = this.resultExpressions.size();
 				if (resultExpressionsCount == 0)
 					return this.resolvedType = null; // error flagging would have been done during the earlier phase.
 				boolean yieldErrors = false;
 				for (int i = 0; i < resultExpressionsCount; i++) {
 					Expression resultExpr = this.resultExpressions.get(i);
-					TypeBinding origType = this.originalTypeMap.get(resultExpr);
-					// NB: if origType == null we assume that initial resolving failed hard, rendering re-resolving impossible
-					if (origType != null &&  origType.kind() == Binding.POLY_TYPE) {
+					if (resultExpr.isPolyExpression()) {
```

Comment:
Including a null check here: `if (resultExpr.resolvedType != null && resultExpr.isPolyExpression()) {` would make it fully equivalent to what is on master/HEAD I think